### DISCUSSION TITLE
Generating definitions from list of strings

### DISCUSSION THREAD

**Discussion by jpluscplusm** [cue-lang/cue#1659] [Q&A]
Hi all 👋 

I've got a list of strings (`x: [ "abc", "def" ]`) coming from an external, ad-hoc JSON-based schema. I'd like to provide users of this schema with some CUE-based definitions, automatically inferred from the imported JSON.  

What syntax lets me iterate over `x` and define Definitions `#abc` and `#def`?

This works if I don't need them to be Definitions:

```
$ cue def
x: ["abc", "def"]
for name in x {
	(name): {
		...
	}
}
$ cue eval
x: ["abc", "def"]
abc: {}
def: {}
```

But I can't figure out how to make them into Definitions. No combination of string interpolation or "just jamming #-signs in there" seems to work :-)

Any ideas?

**Comment by mvdan** [comment]
Perhaps this is a bit of an [XY problem](https://en.wikipedia.org/wiki/XY_problem) situation - what exactly are you trying to accomplish by turning them into definitions?

For example, if you want to hide them from the exported output, you could use `_abc` and `_def`. Or, if you wanted to close the structs just like in any definition, you can do that with the `close` builtin too.

**Reply by jlongtine** [reply]
One subtle distinction re `close`... it doesn't recursively close like `#Def` does. So `close` may not _quite_ suffice for @jpluscplusm's needs (depends on the specifics, obviously).

**Reply by jpluscplusm** [reply]
@mvdan I was looking to provide some top-level, importable definitions based on existing non-CUE schemas that I don't control. I'm still doing that, in a different context, so I'll probably kick off a more targeted discussion :-)

I'm still interested in the underlying question, here, though - but I'm very happy to swim with the tide in terms of how we're expecting CUE to interop with the wide packager/schema/hosting/etc ecosystem!

**Comment by rogpeppe** [comment]
Another possibility: how about producing a single definition containing a member for each of the original definitions?

For example:

```
x: ["abc", "def"]
#s: {
	for name in x {
		(name): {
			...
		}
	}
}
```

This doesn't quite accomplish your original goal of having a top level definition for each original member (I don't think that's possible without doing it outside CUE currently), but I believe it's functionally identical, and each member will be recursively closed as with a regular definition.

**Reply by jpluscplusm** [reply]
Interesting - thanks for the idea, @rogpeppe! This feels close to what I was trying to achieve, and I guess the stylistic oddness of struct-in-def isn't too much for folks to acclimatise to. \<aside> This /kind/ of thing feels like it's something crying out for the Idiomatic CUE docs that were mentioned in the recent Documentation community call! \</aside>
